File: backend/src/main/java/com/petconnect/backend/record/application/service/RecordService.java
package com.petconnect.backend.record.application.service;

import com.petconnect.backend.exception.RecordImmutableException;
import com.petconnect.backend.record.application.dto.*;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;


/**
 * Service interface for managing medical records ({@link Record}) for pets.
 * Defines operations for creating, retrieving, and potentially deleting records.
 * Implementations must handle authorization checks (e.g., ensuring owner/staff access)
 * and business logic (e.g., handling vaccine details, signing records).
 *
 * @author ibosquet
 */
public interface RecordService {
    /**
     * Creates a new medical record for a specific pet.
     * Can be called by the pet's owner or authorized clinic staff.
     * If the type is VACCINE, vaccine details from the DTO must be included and persisted.
     * If created by a Vet, allows for optional immediate signing.
     * @param createDto     The DTO containing the record details (type, description, optional vaccine).
     * @param creatorUserId The ID of the User (Owner, Vet, Admin) creating the record.
     * @return The DTO representation of the newly created record.
     * @throws com.petconnect.backend.exception.EntityNotFoundException if the pet or creator user is not found,
     *                                                                  or if vaccine details are provided, but breed is invalid for a vaccine type.
     * @throws org.springframework.security.access.AccessDeniedException  if the creator is not authorized to create a record for this pet.
     * @throws IllegalArgumentException if the type is VACCINE, but vaccine details are missing in the DTO, or vice versa.
     * @throws IllegalStateException if signing is requested but the creator is not a Vet.
     */
    RecordViewDto createRecord( RecordCreateDto createDto, Long creatorUserId);

    /**
     * Retrieves a paginated list of all medical records for a specific pet.
     * Requires the requester to be the pet's owner or authorized clinic staff.
     * Records are typically ordered by the creation date descending.
     *
     * @param petId           The ID of the pet whose records are to be retrieved.
     * @param requesterUserId The ID of the user requesting the records (for authorization).
     * @param pageable        Pagination and sorting information.
     * @return A Page containing {@link RecordViewDto} objects.
     * @throws com.petconnect.backend.exception.EntityNotFoundException if the pet is not found.
     * @throws org.springframework.security.access.AccessDeniedException  if the requester is not authorized to view records for this pet.
     */
    Page<RecordViewDto> findRecordsByPetId(Long petId, Long requesterUserId, Pageable pageable);

    /**
     * Retrieves a specific medical record by its ID.
     * Requires the requester to be the pet's owner or authorized clinic staff.
     *
     * @param recordId        The ID of the record to retrieve.
     * @param requesterUserId The ID of the user requesting the record (for authorization).
     * @return The {@link RecordViewDto} of the found record.
     * @throws com.petconnect.backend.exception.EntityNotFoundException if the record is not found.
     * @throws org.springframework.security.access.AccessDeniedException  if the requester is not authorized to view this record.
     */
    RecordViewDto findRecordById(Long recordId, Long requesterUserId);

    /**
     * Updates an existing *unsigned* medical record.
     * Allows updating the type (excluding VACCINE) and/or description.
     * Requires the requester to be the original creator (Owner/Staff) or an authorized Admin
     * if the record was created by clinic staff.
     *
     * @param recordId        The ID of the unsigned record to update.
     * @param updateDto       The DTO containing the fields to update (type, description).
     * @param requesterUserId The ID of the user attempting the update.
     * @return The DTO representation of the updated record.
     * @throws com.petconnect.backend.exception.EntityNotFoundException if the record or requester is not found.
     * @throws org.springframework.security.access.AccessDeniedException  if the requester is not authorized to update this record.
     * @throws IllegalStateException if the record is signed or if attempting to change the type to/from VACCINE.
     * @throws IllegalArgumentException if the new type is invalid.
     */
    RecordViewDto updateUnsignedRecord(Long recordId, RecordUpdateDto updateDto, Long requesterUserId);

    /**
     * Deletes a medical record based on its ID and the requester's authorization.
     * - Unsigned records can be deleted by their creator or an authorized Admin.
     * - Signed records can ONLY be deleted by the signing Vet IF the record is NOT immutable.
     *
     * @param recordId        The ID of the record to delete.
     * @param requesterUserId The ID of the user attempting the deletion.
     * @throws EntityNotFoundException if the record or requester user is not found.
     * @throws RecordImmutableException if attempting to delete a signed record linked to a certificate.
     * @throws AccessDeniedException if the requester is not authorized to delete the record.
     */
    void deleteRecord(Long recordId, Long requesterUserId);

    /**
     * Generates a temporary access token (JWT) allowing read-only access
     * to a specific pet's signed medical records for a limited duration.
     * Requires the requester to be the owner of the pet.
     *
     * @param petId           The ID of the pet whose records access is requested.
     * @param requestDto      DTO containing the requested duration.
     * @param requesterUserId The ID of the user making the request (must be the owner).
     * @return A DTO containing the generated temporary access token.
     * @throws EntityNotFoundException if the pet is not found.
     * @throws AccessDeniedException  if the requester is not the owner of the pet.
     * @throws IllegalArgumentException if the duration string in the DTO is invalid.
     */
    TemporaryAccessTokenDto generateTemporaryAccessToken(Long petId, TemporaryAccessRequestDto requestDto, Long requesterUserId);

    /**
     * Retrieves a list of medical records linked to a temporary access token.
     * The token provides limited, read-only access to signed medical records.
     *
     * @param tokenValue The value of the temporary access token used for authorization.
     * @return A list of {@link RecordViewDto} objects representing the medical records accessible with the provided token.
     */
    @Transactional(readOnly = true)
    List<RecordViewDto> findRecordsByTemporaryAccessToken(String tokenValue);

    /**
     * Retrieves a paginated list of all medical records created by staff within a specific clinic.
     * This serves as a historical view for the clinic.
     *
     * @param clinicId The ID of the clinic.
     * @param requesterUserId The ID of the user (clinic staff) requesting the records.
     * @param pageable Pagination information.
     * @return A Page of {@link RecordViewDto}.
     * @throws AccessDeniedException if the requester is not authorized, staff of the clinic.
     */
    Page<RecordViewDto> findRecordsCreatedByClinic(Long clinicId, Long requesterUserId, Pageable pageable);
}

File: frontend/src/services/recordService.ts
import axios from 'axios';
import { API_BASE_URL } from '@/config';
import {
    RecordViewDto,
    RecordCreatePayload,
    RecordUpdatePayload,
    Page,
    ApiErrorResponse,
    RecordType,
    TemporaryAccessRequestPayload,
    TemporaryAccessTokenResponse
} from '@/types/apiTypes'; 

interface FindRecordsParams {
    petId: number | string;
    page: number;
    size: number;
    sort?: string; 
}

interface FindClinicRecordsParams {
    clinicId: number | string;
    page: number;
    size: number;
    sort?: string;
    petName?: string;
    ownerName?: string;
    recordType?: RecordType | "";
    vetCreatorId?: number | string;
}

interface AxiosQueryParameters {
    page: number;
    size: number;
    sort?: string;
    petName?: string;
    ownerName?: string;
    recordType?: string; 
    vetCreatorId?: string;
}

/**
 * Fetches a paginated list of medical records for a specific pet.
 *
 * @param {string} token - The JWT token for authorization.
 * @param {FindRecordsParams} params - Parameters including petId and pagination.
 * @returns {Promise<Page<RecordViewDto>>} A promise resolving to paginated record data.
 * @throws {Error} If fetching fails.
 */
export const findRecordsByPetId = async (
    token: string,
    { petId, page, size, sort = 'createdAt,desc' }: FindRecordsParams
): Promise<Page<RecordViewDto>> => {
    if (!token) throw new Error("Authentication token required.");
    try {
        const response = await axios.get<Page<RecordViewDto>>(`${API_BASE_URL}/records`, {
            headers: { 'Authorization': `Bearer ${token}` },
            params: { petId, page, size, sort }
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            throw new Error(typeof apiError.message === 'string' ? apiError.message : apiError.error || 'Failed to fetch records.');
        }
        throw new Error('Failed to fetch records due to network or unexpected error.');
    }
};

/**
 * Creates a new medical record for a pet.
 *
 * @param {string} token - The JWT token.
 * @param {RecordCreatePayload} payload - The data for the new record.
 * @returns {Promise<RecordViewDto>} A promise resolving to the created record data.
 * @throws {Error} If creation fails.
 */
export const createRecord = async (
    token: string,
    payload: RecordCreatePayload
): Promise<RecordViewDto> => {
    if (!token) throw new Error("Authentication token required.");
    try {
        const response = await axios.post<RecordViewDto>(`${API_BASE_URL}/records`, payload, {
            headers: {
                'Authorization': `Bearer ${token}`,
                'Content-Type': 'application/json',
            },
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            let errorMessage = 'Failed to create record.';
            if (typeof apiError.message === 'string') { errorMessage = apiError.message; }
            else if (typeof apiError.message === 'object' && apiError.message !== null) { errorMessage = Object.values(apiError.message).join(' '); }
            else if (apiError.error) { errorMessage = apiError.error; }
            throw new Error(errorMessage);
        }
        throw new Error('Failed to create record due to network or unexpected error.');
    }
};

/**
 * Updates an existing unsigned medical record.
 * Only the owner can update their own non-vaccine, unsigned records.
 *
 * @param {string} token - The JWT token.
 * @param {number | string} recordId - The ID of the record to update.
 * @param {RecordUpdatePayload} payload - The data to update the record with.
 * @returns {Promise<RecordViewDto>} A promise resolving to the updated record data.
 * @throws {Error} If update fails.
 */
export const updateUnsignedRecord = async (
    token: string,
    recordId: number | string,
    payload: RecordUpdatePayload
): Promise<RecordViewDto> => {
    if (!token) throw new Error("Authentication token required.");
    try {
        const response = await axios.put<RecordViewDto>(`${API_BASE_URL}/records/${recordId}`, payload, {
            headers: {
                'Authorization': `Bearer ${token}`,
                'Content-Type': 'application/json',
            },
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
             let errorMessage = 'Failed to update record.';
            if (typeof apiError.message === 'string') { errorMessage = apiError.message; }
            else if (typeof apiError.message === 'object' && apiError.message !== null) { errorMessage = Object.values(apiError.message).join(' '); }
            else if (apiError.error) { errorMessage = apiError.error; }
            throw new Error(errorMessage);
        }
        throw new Error('Failed to update record due to network or unexpected error.');
    }
};

/**
 * Deletes a medical record.
 * Restrictions apply based on who created it and if it's signed (handled by backend).
 *
 * @param {string} token - The JWT token.
 * @param {number | string} recordId - The ID of the record to delete.
 * @returns {Promise<void>} A promise that resolves when deletion is successful.
 * @throws {Error} If deletion fails.
 */
export const deleteRecord = async (
    token: string,
    recordId: number | string
): Promise<void> => {
    if (!token) throw new Error("Authentication token required.");
    try {
        await axios.delete(`${API_BASE_URL}/records/${recordId}`, {
            headers: { 'Authorization': `Bearer ${token}` },
        });
    } catch (error) {
         if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            throw new Error(typeof apiError.message === 'string' ? apiError.message : apiError.error || 'Failed to delete record.');
        }
        throw new Error('Failed to delete record due to network or unexpected error.');
    }
};

/**
 * Fetches a single medical record by its ID.
 *
 * @param {string} token - The JWT token.
 * @param {number | string} recordId - The ID of the record.
 * @returns {Promise<RecordViewDto>} A promise resolving to the record data.
 * @throws {Error} If fetching fails.
 */
export const findRecordById = async (token: string, recordId: number | string): Promise<RecordViewDto> => {
    if (!token) throw new Error("Authentication token required.");
    try {
        const response = await axios.get<RecordViewDto>(`${API_BASE_URL}/records/${recordId}`, {
            headers: { 'Authorization': `Bearer ${token}` },
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            throw new Error(typeof apiError.message === 'string' ? apiError.message : apiError.error || 'Failed to fetch record details.');
        }
        throw new Error('Failed to fetch record details due to network or unexpected error.');
    }
};

/**
 * Fetches a paginated list of all medical records created by a specific clinic.
 *
 * @param {string} token - The JWT token for authorization.
 * @param {FindClinicRecordsParams} params - Parameters including clinicId and pagination.
 * @returns {Promise<Page<RecordViewDto>>} A promise resolving to paginated record data.
 * @throws {Error} If fetching fails.
 */
export const findRecordsCreatedByClinic = async (
    token: string,
    { clinicId, page, size, sort = 'createdAt,desc', petName, ownerName, recordType, vetCreatorId }: FindClinicRecordsParams 
): Promise<Page<RecordViewDto>> => {
    if (!token) throw new Error("Authentication token required.");
    if (!clinicId) throw new Error("Clinic ID is required.");

    try {
        const queryParams: AxiosQueryParameters = { page, size, sort };
        if (petName) queryParams.petName = petName;
        if (ownerName) queryParams.ownerName = ownerName;
        if (recordType) queryParams.recordType = recordType; 
        if (vetCreatorId) queryParams.vetCreatorId = vetCreatorId.toString();

        const response = await axios.get<Page<RecordViewDto>>(`${API_BASE_URL}/records/clinic/${clinicId}/created-by`, {
            headers: { 'Authorization': `Bearer ${token}` },
            params: queryParams
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            throw new Error(typeof apiError.message === 'string' ? apiError.message : apiError.error || 'Failed to fetch clinic records.');
        }
        throw new Error('Failed to fetch clinic records due to network or unexpected error.');
    }
};

/**
 * Generates a temporary access token for sharing a pet's signed medical records.
 *
 * @param {string} token - The JWT authentication token of the Owner.
 * @param {number | string} petId - The ID of the pet whose records are to be shared.
 * @param {TemporaryAccessRequestPayload} payload - Contains the duration string (e.g., "PT1H", "P1D").
 * @returns {Promise<TemporaryAccessTokenResponse>} A promise resolving to an object containing the temporary access token.
 * @throws {Error} If token generation fails.
 */
export const generateTemporaryRecordAccessToken = async (
    token: string,
    petId: number | string,
    payload: TemporaryAccessRequestPayload
): Promise<TemporaryAccessTokenResponse> => {
    if (!token) throw new Error("Authentication token required.");
    if (!petId) throw new Error("Pet ID is required.");

    try {
        const response = await axios.post<TemporaryAccessTokenResponse>(
            `${API_BASE_URL}/records/${petId}/temporary-access`,
            payload,
            {
                headers: {
                    'Authorization': `Bearer ${token}`,
                    'Content-Type': 'application/json',
                },
            }
        );
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            console.error(`API Generate Temp Access Token (Pet ID: ${petId}) Error:`, apiError);
            let errMsg = 'Failed to generate temporary access token.';
            if (typeof apiError.message === 'string') errMsg = apiError.message;
            else if (apiError.error) errMsg = apiError.error;
            throw new Error(errMsg);
        }
        throw new Error('Failed to generate temporary access token due to network or unexpected error.');
    }
};

/**
 * Fetches signed medical records for a pet using a temporary access token.
 *
 * @param {string} tempToken - The temporary access token obtained from a shared link/QR.
 * @returns {Promise<RecordViewDto[]>} A promise resolving to a list of signed records.
 * @throws {Error} If fetching or token validation fails.
 */
export const getRecordsByTemporaryToken = async (tempToken: string): Promise<RecordViewDto[]> => {
    if (!tempToken) throw new Error("Temporary access token is required.");
    try {
        const response = await axios.get<RecordViewDto[]>(`${API_BASE_URL}/records/verify-temporary-access`, {
            params: { token: tempToken } 
        });
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            const apiError = error.response.data as ApiErrorResponse;
            let errMsg = 'Failed to retrieve records with token.';
            if (error.response.status === 400) errMsg = apiError.message as string || "Invalid or expired token.";
            else if (typeof apiError.message === 'string') errMsg = apiError.message;
            else if (apiError.error) errMsg = apiError.error;
            throw new Error(errMsg);
        }
        throw new Error('Failed to retrieve records due to network or unexpected error.');
    }
};